### PR TITLE
Replace cmake with meson for minimal C++ demo

### PR DESCRIPTION
**Pull request by pmli**
This works on my machine, although I don't really understand what is going on in `meson.build`. It would be good to check if this works on other machines, i.e., running

```bash
pip install -r requirements-ci.txt
make docs
```

If people are ok with this change, the tutorial on using external solvers should be updated.

### CODE REVIEW

**Comment by sdrave** [comment]
So what's the reason for the change? I'd say that cmake is still pretty much standard ..

**Comment by pmli** [comment]
> So what's the reason for the change? I'd say that cmake is still pretty much standard ..

Meson is used in NumPy and SciPy, so it seems to be becoming popular for Python projects. Also, I gave up on trying to make cmake work on my machine (it detects the system Python instead of the one in the virtual environment), but I guess it could still be fixed...

File: docs/source/minimal_cpp_demo/meson.build [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    include_directories: [pybind11_config_ret.stdout().split('-I')[-1].strip()],
+)
+
+python3 = dependency(py3_inst.path())
+
+py3avlw = py3_inst.extension_module(
+    'model',
+    sources: ['model.cc'],
+    dependencies: [python3, pybind11],
+    link_language: 'cpp',
```
Comment (eli-schwartz) [inline docs/source/minimal_cpp_demo/meson.build:23]: link_language overrides the automatic detection, but in this case you only have one source file, which is already cpp, so:
- there is one language for this target, that being cpp
- meson will try to autodetect the best language out of all languages for this target (cpp is preferred over c for targets that have multiple languages)

So this is not strictly necessary.

File: docs/source/minimal_cpp_demo/meson.build [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+)
+
+pymod = import('python')
+py3_inst = pymod.find_installation('python3')
+
+pybind11_config = find_program('pybind11-config')
+pybind11_config_ret = run_command(pybind11_config, ['--includes'], check: true)
+pybind11 = declare_dependency(
+    include_directories: [pybind11_config_ret.stdout().split('-I')[-1].strip()],
+)
```
[the remark this answers is not in the document]
Comment (pmli) [reply docs/source/minimal_cpp_demo/meson.build]: Thank you very much @eli-schwartz, this helps a lot.

If I may ask, it seems that `meson` finds `pybind11` but not `pybind11-config1`:

```
pybind11-config found: NO
Found CMake: /usr/bin/cmake (3.22.1)
Run-time dependency pybind11 found: YES 2.9.2
```

[the remark this answers is not in the document]
Comment (eli-schwartz) [reply docs/source/minimal_cpp_demo/meson.build]: Yes, pybind11 upstream installs 3 different methods:
- a python script called `pybind11-config`
- a pkg-config file called `pybind11.pc`
- a cmake file called `pybind11Config.cmake`

They all report the same information, and meson will try all 3 to make sure it can find the installed pybind11 headers. It will look for the python script before cmake, though.

It should not matter *how* you find pybind11, only that you do find it (possibly with a certain minimum version). Although if you want to specifically require a given method, you can do this:
```meson
dependency('pybind11', method: 'config-tool')
```

I see no real reason to deny the use of a cmake-detected copy of the pybind11 headers, though.

[the remark this answers is not in the document]
Comment (eli-schwartz) [reply docs/source/minimal_cpp_demo/meson.build]: I notice you added pybind11-global to the dependencies. That's what provides the cmake config, and it's not strictly necessary as long as the pybind11-config script is installed to $PATH, which is why meson prefers that script if it can find it.

I'm not sure how `pip` build isolation plays with scripts installed to PATH -- I think some old versions of pip may have been buggy here?

[the remark this answers is not in the document]
Comment (pmli) [reply docs/source/minimal_cpp_demo/meson.build]: Great, thank you.

I guess we could add a note about using a recent version of `pip`, or suggest another way of installing pybind11.

[the remark this answers is not in the document]
Comment (eli-schwartz) [reply docs/source/minimal_cpp_demo/meson.build]: You can create a "subprojects/" directory, and run `meson wrap install pybind11`. It installs the following file:

```ini
# $ cat subprojects/pybind11.wrap
[wrap-file]
directory = pybind11-2.10.4
source_url = https://github.com/pybind/pybind11/archive/refs/tags/v2.10.4.tar.gz
source_filename = pybind11-2.10.4.tar.gz
source_hash = 832e2f309c57da9c1e6d4542dedd34b24e4192ecb4d62f6f4866a737454c9970
patch_filename = pybind11_2.10.4-1_patch.zip
patch_url = https://wrapdb.mesonbuild.com/v2/pybind11_2.10.4-1/get_patch
patch_hash = 9489d0cdc1244078a3108c52b4591a6f07f3dc30ca7299d3a3c42b84fa763396
source_fallback_url = https://github.com/mesonbuild/wrapdb/releases/download/pybind11_2.10.4-1/pybind11-2.10.4.tar.gz
wrapdb_version = 2.10.4-1

[provide]
pybind11 = pybind11_dep
```

If this file is present, then meson will know how to "provide" pybind11 by downloading this tarball, securely checksumming it, and extracting it to get a meson subproject that "builds" pybind11 and makes it available. (Of course, it is header-only, so this subproject just assigns the header directory and returns the result).

See https://mesonbuild.com/Wrap-dependency-system-manual.html for more discussion on how this works.

tl;dr -- vendoring your dependencies is a common approach to "things should just work out of the box". Meson implements a kind of vendoring that:
- supports both getting an existing installed copy, and using the vendored copy
- doesn't require modifying build files to add vendored copies
- doesn't require checking lots of third-party source code into git
- provides automatic fine-grained control in case you would like to choose whether to use system or vendored copies
